### PR TITLE
#19608 Linearize curved geometry

### LINKED ISSUE #19608: [PostGIS] Support of curves geometries (SQL/MM Part 3) in map viewer
**Is your feature request related to a problem? Please describe.**

I really like the possibility to have an overview of PostGIS data in the map viewer, this is an amazing feature that I couldn't live without. However, curves geometries are currently not supported.

**Describe the solution you'd like**

Support of curves geometries (SQL/MM Part 3) in map viewer.

**Additional context**

- [SQL/MM Part 3](https://postgis.net/docs/using_postgis_dbmanagement.html#SQL_MM_Part3)

### PR DESCRIPTION
**Pull request by ShadelessFox**
This PR features code from the [GeoTools](https://github.com/geotools/geotools) library.

I'm a bit worried about the license as it uses LGPL-2.1. Please let me know what I should do about it.

Snippet for testing:
```sql
select
	st_geomfromtext('CIRCULARSTRING(0 0, 1 1, 1 0)') actual,
	st_curvetoline(st_geomfromtext('CIRCULARSTRING(0 0, 1 1, 1 0)')) expected
union all
select
	st_geomfromtext('CIRCULARSTRING(0 0, 4 0, 4 8, 0 4, 0 0)'),
	st_curvetoline(st_geomfromtext('CIRCULARSTRING(0 0, 4 0, 4 8, 0 4, 0 0)'))
union all
select
	st_geomfromtext('COMPOUNDCURVE( CIRCULARSTRING(0 0, 1 1, 1 0),(1 0, 0 1))'),
	st_curvetoline(st_geomfromtext('COMPOUNDCURVE( CIRCULARSTRING(0 0, 1 1, 1 0),(1 0, 0 1))'))
union all
select
	st_geomfromtext('CURVEPOLYGON( CIRCULARSTRING(0 0, 4 0, 4 4, 0 4, 0 0), (1 1, 3 3, 3 1, 1 1) )'),
	st_curvetoline(st_geomfromtext('CURVEPOLYGON( CIRCULARSTRING(0 0, 4 0, 4 4, 0 4, 0 0), (1 1, 3 3, 3 1, 1 1) )'))
union all
select
	st_geomfromtext('CURVEPOLYGON( COMPOUNDCURVE( CIRCULARSTRING(0 0,2 0, 2 1, 2 3, 4 3), (4 3, 4 5, 1 4, 0 0)), CIRCULARSTRING(1.7 1, 1.4 0.4, 1.6 0.4, 1.6 0.5, 1.7 1) )'),
	st_curvetoline(st_geomfromtext('CURVEPOLYGON( COMPOUNDCURVE( CIRCULARSTRING(0 0,2 0, 2 1, 2 3, 4 3), (4 3, 4 5, 1 4, 0 0)), CIRCULARSTRING(1.7 1, 1.4 0.4, 1.6 0.4, 1.6 0.5, 1.7 1) )'))
union all
select
	st_geomfromtext('MULTICURVE( (0 0, 5 5), CIRCULARSTRING(4 0, 4 4, 8 4))'),
	st_curvetoline(st_geomfromtext('MULTICURVE( (0 0, 5 5), CIRCULARSTRING(4 0, 4 4, 8 4))'))
union all
select
	st_geomfromtext('MULTISURFACE( CURVEPOLYGON( CIRCULARSTRING( 0 0, 4 0, 4 4, 0 4, 0 0), (1 1, 3 3, 3 1, 1 1)), ((10 10, 14 12, 11 10, 10 10), (11 11, 11.5 11, 11 11.5, 11 11)))'),
	st_curvetoline(st_geomfromtext('MULTISURFACE( CURVEPOLYGON( CIRCULARSTRING( 0 0, 4 0, 4 4, 0 4, 0 0), (1 1, 3 3, 3 1, 1 1)), ((10 10, 14 12, 11 10, 10 10), (11 11, 11.5 11, 11 11.5, 11 11)))'));
```

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java [hunk 1]
```diff
@@ -67,4 +74,94 @@ public static DBGeometry parseWKB(String hexString) throws DBCException {
...
+        return value instanceof CircularString
+            || value instanceof CompoundCurve
+            || value instanceof CurvePolygon
+            || value instanceof MultiCurve
+            || value instanceof MultiSurface;
+    }
+
+    @NotNull
+    public static Object linearize(@NotNull Geometry value) {
+        return linearize(value, 1 / 32.0);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java]: Why 32? Can we add a comment or make a constant?

File: plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/model/gis/CircularArc.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+import org.jkiss.code.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Taken from the GeoTools library
+ *
+ * @see <a href="https://github.com/geotools/geotools/blob/main/modules/library/main/src/main/java/org/geotools/geometry/jts/CircularString.java">CircularString.java</a>
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/model/gis/CircularArc.java]: I don't get it. Did you take this whole class or make your own by example of CircularString?

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/model/gis/CircularArc.java]: Oops, wrong link. Thank you.

File: plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java (hunk 1 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java]: The `ST_CurveToLine` function that converts curved geometry to line geometry uses this value to produce 32 segments per quadrant, so I took it as a reference. I'll add a comment.

File: plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java [hunk 3]
```diff
@@ -84,7 +84,8 @@ public static boolean isCurve(@Nullable Object value) {
 
     @NotNull
     public static Object linearize(@NotNull Geometry value) {
-        return linearize(value, 1 / 32.0);
+        // This value results in 32 segments per quadrant, the default tolerance for ST_CurveToLine
+        return linearize(value, 0.001);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java:88]: 1/32.0 is not equals to 0.001. Is it a mistake?

Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.data.gis/src/org/jkiss/dbeaver/data/gis/handlers/WKGUtils.java:88]: It's not. `1/32` was my crude assumption on how the number of segments is computed, and it was wrong.

**Review by E1izabeth — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**